src/System.CommandLine/Help/HelpBuilder.cs

```diff
@@ -204,103 +206,199 @@ private void AppendDescription(string description)
-        /// Maximum number of characters accross all <see cref="HelpItem">help items</see>
-        /// occupied by the invocation text
+        /// <param name="helpItems">
+        /// Collection of <see cref="HelpItem"/> to write to the console.
         /// </param>
-        protected void AppendHelpItem(
-            HelpItem helpItem, 
-            int maxInvocationWidth)
+        /// <exception cref="ArgumentNullException"></exception>
+        protected virtual void AppendHelpItems(IReadOnlyCollection<HelpItem> helpItems)
         {
-            if (helpItem is null)
-            {
-                throw new ArgumentNullException(nameof(helpItem));
-            }
-
-            AppendText(helpItem.Invocation, CurrentIndentation);
+            if (helpItems is null)
+                throw new ArgumentNullException(nameof(helpItems));
 
-            var offset = maxInvocationWidth + ColumnGutter - helpItem.Invocation.Length;
-            var availableWidth = GetAvailableWidth();
-            var maxDescriptionWidth = availableWidth - maxInvocationWidth - ColumnGutter;
-            var descriptionColumn = helpItem.Description;
-            if (helpItem.HasDefaultValueHint)
+            var table = CreateTable(helpItems, item => new[]
             {
-                var postfix = string.IsNullOrEmpty(helpItem.Description) ? string.Empty : " ";
-                descriptionColumn += postfix + helpItem.DefaultValueHint;
-            }
-            var descriptionLines = SplitText(descriptionColumn, maxDescriptionWidth);
-            var lineCount = descriptionLines.Count;
+                item.Invocation,
+                JoinNonEmpty(" ", item.Description, item.DefaultValueHint)
+            });
 
-            AppendLine(descriptionLines.FirstOrDefault(), offset);
+            var columnWidths = ColumnWidths(table);
+            AppendTable(table, columnWidths);
+        }
 
-            if (lineCount == 1)
-            {
-                return;
+        /// <summary>
+        /// Create a read only table of strings using the projection of a collection.
+        /// </summary>
+        /// <typeparam name="T">The type of the elements of <paramref name="collection"/>.</typeparam>
+        /// <param name="collection">The collection of values to create the table from.</param>
+        /// <param name="selector">A transformation function to apply to each element of <paramref name="collection"/>.</param>
+        /// <returns>
+        /// A read only table of strings whose elements are the projection of the collection with whitespace formatting removed.
+        /// </returns>
+        protected virtual ReadOnlyCollection<ReadOnlyCollection<string>> CreateTable<T>(IEnumerable<T> collection, Func<T, IEnumerable<string>> selector)
+        {
+            return collection.Select(selector)
+                .Select(row => row
+                    .Select(element => RemoveFormatting(element))
+                    .ToList().AsReadOnly())
```

Comment:
Is this needed? Does the the outer `ToList` suffice?

Reply:
A ReadOnlyCollection is not necessary but I used one to follow the [guidelines for collections](https://docs.microsoft.com/en-us/dotnet/standard/design-guidelines/guidelines-for-collections).

I think a multidimensional array is better suited to represent a table. Originally I was going to implement the table view as a separate class with a multidimensional array backing but decided to implement it in place instead so I opted for a ReadOnlyCollection.

Reply:
I believe that guidance predates the introduction of `IReadOnlyCollection<T>` and `IReadOnlyList<T>`. Generally, it's good to return a less specific type than a more specific type.

I'd avoid the multidimensional array as it's harder to reason about.

We have a table abstraction in System.CommandLine.Rendering. That project was started in part out of a desire to have better and more flexible help layouts. I wonder if there's an opportunity for reuse or convergence there. 

@Keboo?